File: eeshaan/algorithm.py
import sys
from collections import deque

# ─── Maze constants ──────────────────────────────────────────
MAZE_SIZE = 16

# ─── Direction constants ─────────────────────────────────────
NORTH = 0
EAST  = 1
SOUTH = 2
WEST  = 3

# ─── Robot state ─────────────────────────────────────────────
robot_x   = 0
robot_y   = 0
robot_dir = NORTH

# ─── Maze data ───────────────────────────────────────────────
walls   = [[[False] * 4 for _ in range(MAZE_SIZE)] for _ in range(MAZE_SIZE)]
visited = [[False]      * MAZE_SIZE                for _ in range(MAZE_SIZE)]

# ─── MMS API ─────────────────────────────────────────────────
def move_forward():
    print("moveForward", flush=True)
    input()

def turn_left():
    print("turnLeft", flush=True)
    input()

def turn_right():
    print("turnRight", flush=True)
    input()

def wall_front():
    print("wallFront", flush=True)
    return input() == "true"

def wall_left():
    print("wallLeft", flush=True)
    return input() == "true"

def wall_right():
    print("wallRight", flush=True)
    return input() == "true"

def log(text):
    print(f"log {text}", flush=True)

def set_color(x, y, color):
    print(f"setColor {x} {y} {color}", flush=True)

def set_text(x, y, text):
    print(f"setText {x} {y} {text}", flush=True)

# ─── Helpers ─────────────────────────────────────────────────
def is_goal(x, y):
    return x in [7, 8] and y in [7, 8]

def in_bounds(x, y):
    return 0 <= x < MAZE_SIZE and 0 <= y < MAZE_SIZE

def neighbour(x, y, direction):
    if direction == NORTH: return x,     y + 1
    if direction == EAST:  return x + 1, y
    if direction == SOUTH: return x,     y - 1
    if direction == WEST:  return x - 1, y

# ─── Sense walls around current cell ─────────────────────────
def sense_walls():
    front = wall_front()
    left  = wall_left()
    right = wall_right()

    dir_front = robot_dir
    dir_left  = (robot_dir - 1) % 4
    dir_right = (robot_dir + 1) % 4

    x, y = robot_x, robot_y

    for has_wall, direction in [(front, dir_front), (left, dir_left), (right, dir_right)]:
        if has_wall:
            walls[x][y][direction] = True
            nx, ny = neighbour(x, y, direction)
            if in_bounds(nx, ny):
                walls[nx][ny][(direction + 2) % 4] = True  # shared wall

# ─── Turn to face a direction ─────────────────────────────────
def face_direction(target_dir):
    global robot_dir
    turns = (target_dir - robot_dir) % 4
    if turns == 1:
        turn_right()
    elif turns == 2:
        turn_right()
        turn_right()
    elif turns == 3:
        turn_left()
    robot_dir = target_dir

# ─── Update position after moving ────────────────────────────
def update_position():
    global robot_x, robot_y
    robot_x, robot_y = neighbour(robot_x, robot_y, robot_dir)
    visited[robot_x][robot_y] = True
    set_color(robot_x, robot_y, "green")

# ─── BFS to find shortest path using known walls ─────────────
# This is the core BFS — finds path from current pos to goal
# Returns list of directions to follow
def bfs_find_path():
    start = (robot_x, robot_y)

    # each item in queue = (x, y, path taken to get here)
    queue   = deque()
    queue.append((start[0], start[1], []))

    seen = [[False] * MAZE_SIZE for _ in range(MAZE_SIZE)]
    seen[start[0]][start[1]] = True

    while queue:
        x, y, path = queue.popleft()

        # reached goal?
        if is_goal(x, y):
            return path

        # explore all 4 neighbours
        for direction in [NORTH, EAST, SOUTH, WEST]:
            if walls[x][y][direction]:
                continue  # wall here, skip

            nx, ny = neighbour(x, y, direction)

            if not in_bounds(nx, ny):
                continue  # out of bounds, skip

            if seen[nx][ny]:
                continue  # already visited in this BFS, skip

            seen[nx][ny] = True
            queue.append((nx, ny, path + [direction]))  # add direction to path

    return []  # no path found

# ─── Exploration phase — discover walls by moving around ──────
# Uses simple left hand rule to explore until goal found
def explore():
    log("Phase 1 — Exploring maze...")

    while not is_goal(robot_x, robot_y):
        sense_walls()

        x, y = robot_x, robot_y

        # try left → front → right → back (left hand rule)
        dir_left  = (robot_dir - 1) % 4
        dir_front = robot_dir
        dir_right = (robot_dir + 1) % 4
        dir_back  = (robot_dir + 2) % 4

        moved = False
        for direction in [dir_left, dir_front, dir_right, dir_back]:
            if not walls[x][y][direction]:
                nx, ny = neighbour(x, y, direction)
                if in_bounds(nx, ny):
                    face_direction(direction)
                    move_forward()
                    update_position()
                    moved = True
                    break

        if not moved:
            log("Stuck! No moves available")
            break

    log("Goal found during exploration!")

# ─── Speed run phase — follow BFS shortest path ───────────────
def speed_run(path):
    log(f"Phase 2 — Running shortest path ({len(path)} steps)...")

    # color the planned path yellow before running
    x, y = 0, 0
    for direction in path:
        nx, ny = neighbour(x, y, direction)
        set_color(nx, ny, "yellow")
        x, y = nx, ny

    # now actually follow the path
    for direction in path:
        face_direction(direction)
        move_forward()
        update_position()

    log("Speed run complete!")

# ─── Main ─────────────────────────────────────────────────────
def solve():
    global robot_x, robot_y, robot_dir

    # mark start cell
    visited[0][0] = True
    set_color(0, 0, "green")

    # Phase 1 — explore until goal is reached
    explore()

    # reset to start
    log("Resetting to start for speed run...")
    robot_x   = 0
    robot_y   = 0
    robot_dir = NORTH

    # Phase 2 — BFS on discovered walls to find shortest path
    path = bfs_find_path()

    if not path:
        log("No path found after exploration!")
        return

    log(f"BFS found path of length {len(path)}")

    # Phase 3 — run the shortest path
    speed_run(path)

    log("Done!")

solve()

File: code/Algorithm.cpp
#include "Algorithm.h"
#include "API.h"
#include "Maze.h"
#include <queue>
#include <vector>
#include <string.h>

// ===========================================================================
// LOGGING  (Serial replaces Python's stderr)
// ===========================================================================
#define LOG(msg)      Serial.println(msg)
#define LOGF(...)     Serial.printf(__VA_ARGS__)

// ===========================================================================
// MOVEMENT HELPERS
// Exact translation of Python turn_left / turn_right / move_forward /
// turn_to_face / move_to_cell.
// ===========================================================================

static void doTurnLeft(MouseState& s) {
    turnLeft();
    s.facing = (s.facing + 3) % 4;   // NORTH→WEST→SOUTH→EAST
}

static void doTurnRight(MouseState& s) {
    turnRight();
    s.facing = (s.facing + 1) % 4;   // NORTH→EAST→SOUTH→WEST
}

static void doMoveForward(MouseState& s) {
    moveForward();
    s.x += DX[s.facing];
    s.y += DY[s.facing];
}

static void turnToFace(uint8_t targetDir, MouseState& s) {
    int rightTurns = (targetDir - s.facing + 4) % 4;
    if      (rightTurns == 1) doTurnRight(s);
    else if (rightTurns == 2) { doTurnRight(s); doTurnRight(s); }  // U-turn
    else if (rightTurns == 3) doTurnLeft(s);
    // 0 → already facing the right way
}

static void moveToCell(uint8_t nx, uint8_t ny, MouseState& s) {
    int8_t dx = (int8_t)nx - (int8_t)s.x;
    int8_t dy = (int8_t)ny - (int8_t)s.y;
    uint8_t targetDir;
    if      (dx ==  1) targetDir = EAST;
    else if (dx == -1) targetDir = WEST;
    else if (dy ==  1) targetDir = NORTH;
    else               targetDir = SOUTH;
    turnToFace(targetDir, s);
    doMoveForward(s);
}

// ===========================================================================
// WALL SENSING
// Converts relative sensor readings (front/left/right) to absolute
// compass directions and records any newly discovered walls.
// Returns true if at least one new wall was found.
// ===========================================================================
static uint8_t absDir(uint8_t facing, uint8_t relative) {
    // relative: 0=front, 1=right, 2=left  (matches sensor call order below)
    if (relative == 0) return facing;
    if (relative == 1) return (facing + 1) % 4;   // right
    return                     (facing + 3) % 4;   // left
}

static bool senseAndRecord(MouseState& s) {
    bool anyNew = false;
    // {relative_index, sensor_reading}
    bool readings[3] = { wallFront(), wallRight(), wallLeft() };
    uint8_t relDirs[3] = { 0, 1, 2 };   // front, right, left
    for (int i = 0; i < 3; i++) {
        if (readings[i]) {
            uint8_t dir = absDir(s.facing, relDirs[i]);
            if (recordWall(s.x, s.y, dir)) anyNew = true;
        }
    }
    return anyNew;
}

// ===========================================================================
// FLOODFILL PHASE
// Navigates from current position to any cell in targets[] by always
// stepping to the open neighbour with the lowest live flood value.
// Senses walls at every cell; rebuilds flood table on any new wall.
// Returns true if a target was reached.
// ===========================================================================
static bool floodfillPhase(MouseState& state,
                            const uint8_t targX[], const uint8_t targY[],
                            uint8_t targCount,
                            const char* label)
{
    // Working flood table for this phase
    int16_t live[MAZE_WIDTH][MAZE_HEIGHT];
    buildLiveFlood(targX, targY, targCount, live);

    int steps = 0;

    // Check if already at a target
    auto atTarget = [&]() {
        for (int i = 0; i < targCount; i++)
            if (state.x == targX[i] && state.y == targY[i]) return true;
        return false;
    };

    while (!atTarget()) {
        uint8_t x = state.x, y = state.y;

        // Sense walls; rebuild if anything new
        bool anyNew = false;
        bool readings[3] = { wallFront(), wallRight(), wallLeft() };
        uint8_t relDirs[3] = { 0, 1, 2 };
        for (int i = 0; i < 3; i++) {
            if (readings[i]) {
                uint8_t dir = absDir(state.facing, relDirs[i]);
                if (recordWall(x, y, dir)) anyNew = true;
            }
        }
        if (anyNew) buildLiveFlood(targX, targY, targCount, live);

        // Pick open neighbour with lowest flood distance
        uint8_t bestDir = 255;
        int16_t bestVal = INF_DIST;
        for (int d = 0; d < 4; d++) {
            if (hasWall(x, y, d)) continue;
            int8_t nx = x + DX[d];
            int8_t ny = y + DY[d];
            if (nx < 0 || nx >= MAZE_WIDTH || ny < 0 || ny >= MAZE_HEIGHT) continue;
            if (live[nx][ny] < bestVal) {
                bestVal = live[nx][ny];
                bestDir = d;
            }
        }

        if (bestDir == 255 || bestVal == INF_DIST) {
            LOGF("  %s: trapped at (%d,%d)!\n", label, x, y);
            return false;
        }

        uint8_t nx = x + DX[bestDir];
        uint8_t ny = y + DY[bestDir];
        moveToCell(nx, ny, state);
        steps++;

        if (!visited[nx][ny]) {
            visited[nx][ny] = true;
        }
    }

    LOGF("  %s: reached target in %d steps\n", label, steps);
    return true;
}

// ===========================================================================
// RUN 1 — FLOODFILL EXPLORATION
// Phase 1: start → goal
// Phase 2: goal  → start
// ===========================================================================
void floodfillExplore(MouseState& state) {
    LOG("=== RUN 1: Floodfill Explore (start->goal->start) ===");

    visited[state.x][state.y] = true;

    // Phase 1: navigate to goal
    LOG("  Phase 1: start -> goal...");
    bool ok = floodfillPhase(state, GOAL_X, GOAL_Y, NUM_GOALS, "Phase1");
    if (!ok) { LOG("  Could not reach goal. Maze disconnected?"); return; }
    LOGF("  Goal reached at (%d,%d). Coverage: %.1f%%\n",
         state.x, state.y, coverage());

    // Phase 2: navigate back to start
    LOG("  Phase 2: goal -> start...");
    const uint8_t startX[1] = {0};
    const uint8_t startY[1] = {0};
    ok = floodfillPhase(state, startX, startY, 1, "Phase2");
    if (!ok) { LOG("  Could not return to start."); return; }

    LOG("=== FLOODFILL EXPLORATION COMPLETE ===");
    LOGF("    Coverage: %.1f%%  Mouse at (%d,%d)\n", coverage(), state.x, state.y);
}

// ===========================================================================
// A* PLANNER
// Plans shortest path from (sx,sy) to any goal cell using flood_dist as
// the heuristic h(n).  Returns path as vector of {x,y} pairs, start
// exclusive, goal inclusive.  Empty vector = no path found.
//
// C++ translation notes:
//   Python heapq    → std::priority_queue (min-heap via greater<>)
//   Python dict     → flat arrays indexed by x*HEIGHT+y (256 cells max)
//   Python set      → bool closed[16][16]
// ===========================================================================
struct AStarNode {
    int16_t f, g;
    uint8_t x, y;
    bool operator>(const AStarNode& o) const { return f > o.f; }
};

struct Cell { uint8_t x, y; };

static std::vector<Cell> astarPlan(uint8_t sx, uint8_t sy) {
    // g_score and came_from indexed as [x][y]
    int16_t g_score[MAZE_WIDTH][MAZE_HEIGHT];
    uint8_t from_x[MAZE_WIDTH][MAZE_HEIGHT];
    uint8_t from_y[MAZE_WIDTH][MAZE_HEIGHT];
    bool    closed[MAZE_WIDTH][MAZE_HEIGHT];

    for (int x = 0; x < MAZE_WIDTH; x++)
        for (int y = 0; y < MAZE_HEIGHT; y++) {
            g_score[x][y] = INF_DIST;
            closed[x][y]  = false;
            from_x[x][y]  = 255;   // sentinel = no parent
            from_y[x][y]  = 255;
        }

    g_score[sx][sy] = 0;

    std::priority_queue<AStarNode, std::vector<AStarNode>, std::greater<AStarNode>> open;
    open.push({flood_dist[sx][sy], 0, sx, sy});

    while (!open.empty()) {
        AStarNode cur = open.top(); open.pop();
        uint8_t x = cur.x, y = cur.y;

        if (closed[x][y]) continue;
        closed[x][y] = true;

        // Goal reached — reconstruct path
        if (isGoal(x, y)) {
            std::vector<Cell> path;
            uint8_t cx = x, cy = y;
            while (!(cx == sx && cy == sy)) {
                path.push_back({cx, cy});
                uint8_t px = from_x[cx][cy];
                uint8_t py = from_y[cx][cy];
                cx = px; cy = py;
            }
            // Reverse so path goes from start to goal
            for (int i = 0, j = path.size()-1; i < j; i++, j--)
                std::swap(path[i], path[j]);
            LOGF("  A* path found: %d steps\n", (int)path.size());
            return path;
        }

        // Expand neighbours
        for (int d = 0; d < 4; d++) {
            if (hasWall(x, y, d)) continue;
            int8_t nx = x + DX[d];
            int8_t ny = y + DY[d];
            if (nx < 0 || nx >= MAZE_WIDTH || ny < 0 || ny >= MAZE_HEIGHT) continue;
            if (closed[nx][ny]) continue;

            int16_t new_g = g_score[x][y] + 1;
            if (new_g < g_score[nx][ny]) {
                g_score[nx][ny] = new_g;
                from_x[nx][ny]  = x;
                from_y[nx][ny]  = y;
                int16_t f = new_g + flood_dist[nx][ny];
                open.push({f, new_g, (uint8_t)nx, (uint8_t)ny});
            }
        }
    }

    LOG("  A* found no path!");
    return {};
}

// ===========================================================================
// RUN 2 — A* SPEED RUN WITH LIVE REPLANNING
//
// Plans the full route, then follows it step by step.
// Before every move, senses walls. If a new wall blocks the planned path,
// rebuilds flood_dist and replans from the current position.
// This prevents any crash from a wall missed during exploration.
// ===========================================================================
void astarRun(MouseState& state) {
    LOG("=== RUN 2: A* Speed Run (flood-fill heuristic, live replanning) ===");

    // Plan initial path
    LOG("  Planning initial path...");
    std::vector<Cell> path = astarPlan(state.x, state.y);
    if (path.empty()) { LOG("  No path found. Aborting."); return; }
    LOGF("  Path: %d steps. Starting run...\n", (int)path.size());

    // Build a quick lookup: is cell (x,y) on the current planned path?
    // Use a bool grid — O(1) lookup, no heap allocation per step.
    bool pathSet[MAZE_WIDTH][MAZE_HEIGHT];
    auto rebuildPathSet = [&]() {
        memset(pathSet, 0, sizeof(pathSet));
        for (auto& c : path) pathSet[c.x][c.y] = true;
    };
    rebuildPathSet();

    int totalSteps = 0;

    while (!isGoal(state.x, state.y)) {
        if (path.empty()) {
            LOG("  Path exhausted — replanning...");
            path = astarPlan(state.x, state.y);
            if (path.empty()) { LOG("  No path. Aborting."); return; }
            rebuildPathSet();
        }

        // --- Sense walls BEFORE moving ---
        uint8_t x = state.x, y = state.y;
        bool anyNew = false;
        bool readings[3] = { wallFront(), wallRight(), wallLeft() };
        uint8_t relDirs[3] = { 0, 1, 2 };
        for (int i = 0; i < 3; i++) {
            if (readings[i]) {
                uint8_t dir = absDir(state.facing, relDirs[i]);
                if (recordWall(x, y, dir)) anyNew = true;
            }
        }

        if (anyNew) {
            // Check if any newly blocked neighbour is on the planned path
            bool blockedOnPath = false;
            for (int d = 0; d < 4; d++) {
                if (!hasWall(x, y, d)) continue;
                int8_t nx = x + DX[d];
                int8_t ny = y + DY[d];
                if (nx < 0 || nx >= MAZE_WIDTH || ny < 0 || ny >= MAZE_HEIGHT) continue;
                if (pathSet[nx][ny]) { blockedOnPath = true; break; }
            }
            if (blockedOnPath) {
                LOGF("  New wall near (%d,%d) blocks path — replanning...\n", x, y);
                buildFloodFill();   // refresh heuristic with updated wall map
                path = astarPlan(x, y);
                if (path.empty()) { LOG("  No path after replan. Aborting."); return; }
                rebuildPathSet();
                LOGF("  Replanned: %d steps remaining.\n", (int)path.size());
            }
        }

        // --- Move to next waypoint ---
        Cell next = path.front();
        path.erase(path.begin());
        moveToCell(next.x, next.y, state);
        totalSteps++;

        if (isGoal(state.x, state.y)) {
            LOGF("  Goal reached at (%d,%d) in %d steps!\n",
                 state.x, state.y, totalSteps);
        }
    }

    LOG("=== A* RUN COMPLETE ===");
}

// ===========================================================================
// RESET HANDLER
// Preserves walls[][] and flood_dist[][] — keeps everything learned.
// Resets only logical position so Run 2 starts from (0,0).
// ===========================================================================
void handleReset(MouseState& state) {
    LOG("Reset detected. Preserving wall map.");
    ackReset();
    state.x      = 0;
    state.y      = 0;
    state.facing = NORTH;
}